### ISSUE TITLE
Deconvolution2D fail to infer shape

### ISSUE THREAD

**Issue by shahurik** [fchollet/keras#3339]
I try to use Deconvolution2D layer with zero initialized weights and get infer_shape error

WARNING (theano.tensor.opt): Failed to infer_shape from Op AbstractConv2d_gradInputs{border_mode='valid', subsample=(1, 1), filter_flip=True, imshp=(1,
 7, 7), kshp=(1, 1, 3, 3), filter_dilation=(1, 1)}.

Full error log is here: https://gist.github.com/shahurik/efa88e378e508c1de658151a5cdefc03
Keras and Theano are upgraded from github just before filling that issue.

Here is minimal code:

``` python
from keras.models import Model
from keras.layers import Input, Convolution2D, Deconvolution2D

from numpy import eye, array

img = Input(shape=(1, 5, 5))
conv = Deconvolution2D(1, 3, 3, output_shape=(1, 7, 7),  init='zero')(img)

model = Model(img, conv)  
model.summary()  
print(model.predict(array([[eye(5)]])))
```

If I change Deconvolution2D to Convolution2D (and delete output_shape argument), works fine.

**Comment by McVilla** [comment]
Hi @shahurik , what version of keras you were using? In keras 1.0.0 or above, I can't import Deconvolution2d.

**Comment by shahurik** [comment]
Hi @McVilla ,
I'm using master branch from github (pip install git+https://github.com/fchollet/keras.git)
Layer is defined here: https://github.com/fchollet/keras/blob/master/keras/layers/convolutional.py#L382

**Comment by McVilla** [comment]
Thanks @shahurik . btw, did you solve your issue?

**Comment by EderSantana** [comment]
Deconv does not do shape inference because of the way it is implemented in Theano and TF. It is a hack arround the conv gradient. We are still looking into how to solve this issue

**Comment by McVilla** [comment]
Thanks, @EderSantana, can I use UpSampling2d and Convolution2d layers to replace deconvolution2d layer?

**Comment by EderSantana** [comment]
it is not quite the same thing. UpSampling2D does pixel repetition when upsampling and I think that deconv puts zeros between pixels to upsample. but I used Up+conv before and it was ok for some tasks, for others it was not... You may have to test
